sentry_streams/sentry_streams/adapters/arroyo/steps.py

```diff
@@ -126,28 +124,63 @@ def transformer(
 
 
 @dataclass
-class StreamSinkStep(ArroyoStep):
+class RouterStep(ArroyoStep):
     """
-    Represents an Arroyo producer. This is mapped from a StreamSink in the pipeline.
+    Represents a Router which can direct a message to one of multiple
+    downstream branches based on the output of a routing function.
 
-    It filters out messages for the route not specified on this step and unpacks
-    the routed message into the original Arroyo payload.
+    Since Arroyo has no concept of 'branches', this updates the `waypoints` list within
+    a message's `Route` object based on the result of the routing function.
     """
 
-    producer: Producer[Any]
-    topic_name: str
+    # TODO: fix generic typing here
+    pipeline_step: Router[RoutingFuncReturnType]  # type: ignore
 
     def build(
-        self, next: ProcessingStrategy[Union[FilteredPayload, RoutedValue]]
+        self, next: ProcessingStrategy[Union[FilteredPayload, RoutedValue]], commit: Commit
     ) -> ProcessingStrategy[Union[FilteredPayload, RoutedValue]]:
-        def extract_value(message: Message[Union[FilteredPayload, RoutedValue]]) -> Any:
-            message_payload = message.value.payload
-            if isinstance(message_payload, RoutedValue) and message_payload.route == self.route:
-                return KafkaPayload(None, str(message_payload.payload).encode("utf-8"), [])
-            else:
-                return FilteredPayload()
+        def append_branch_to_waypoints(
+            payload: RoutedValue,
+        ) -> RoutedValue:
+            routing_func = self.pipeline_step.routing_function
+            routing_table = self.pipeline_step.routing_table
+            result_branch = routing_func(payload.payload)
+            result_branch_name = routing_table[result_branch].name
+            payload.route.waypoints.append(result_branch_name)
+            return payload
+
+        def transformer(
+            message: Message[Union[FilteredPayload, RoutedValue]],
+        ) -> Union[FilteredPayload, RoutedValue]:
+            return process_message(
+                self.route,
+                message,
+                append_branch_to_waypoints,
+            )
```

Comment:
Nit: this may be more concise if you inline it in a lambda in the call to RunTask:
```
return RunTask(
            lambda msg: process_message(
                self.route,
                message,
                append_branch_to_waypoints,
            ),
            next,
        )
```